Avoid exceptions during projection of primary nodes when nodes are aligned

lindsayad opened the pull request:
@recuero added a great comment about loosening the tolerance when
projecting the secondary nodes relative to the primary nodes in
order to avoid our exception about "we should have already registered
this primary node as mapped" during primary node projection for aligned
nodes. However, we appear to have only applied it to the inner if
conditional and missed applying it to the outer one which leads to
exceptions in the input provided in https://github.com/idaholab/moose/discussions/30433

Closes #31308

lindsayad committed:
mpai60 initial contact input

lindsayad committed:
Run formatter

lindsayad committed:
We have to multiply xi tolerance in the outer if check too

Antonio added a great comment about loosening the tolerance when
projecting the secondary nodes relative to the primary nodes in
order to avoid our exception about "we should have already registered
this primary node as mapped" during primary node projection for aligned
nodes. However, we appear to have only applied it to the inner if
conditional and missed applying it to the outer one which leads to
exceptions in the input provided in #30433

lindsayad committed:
Loosen tolerance with mortar mesh changes

lindsayad committed:
Improve exception message

lindsayad committed:
Input file aesthetic changes

lindsayad committed:
Error if user sets penalty for LM mortar

lindsayad committed:
Set eigenstrain value that closes all three cracks

lindsayad committed:
Use correct_edge_dropping

Otherwise you'll see penetration at the boundary edges. This
requires slightly loosenening the relative tolerance

lindsayad committed:
Eliminate rigid body modes

lindsayad committed:
Add test

lindsayad committed:
Replace use of method in test specs

lindsayad committed:
Fix tests now failing bc of missing aux vars

lindsayad committed:
Don't add penetration, nodal area, nor contact pressure when using mortar

lindsayad committed:
Move input into test directory

lindsayad committed:
Fix mortar restart test

The gold file for the restarted case was wrong. This is because
user objects executed on initial destroy stateful material property
values during restart

lindsayad committed:
Fix heavy tests

lindsayad committed:
Loosen c_normal for more robust convergence

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3136315/), step Generate coverage on ba5ae53 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fc4b0</th>
      <th colspan=4>#31268 ba5ae5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.59%</td>
      <td>85.59%</td>
      <td>-0.00%</td>
      <td>60.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>116490</td>
      <td>116491</td>
      <td>+1</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19613</td>
      <td>19614</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fc4b0</th>
      <th colspan=4>#31268 ba5ae5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.81%</td>
      <td>90.62%</td>
      <td>-0.20%</td>
      <td>97.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4903</td>
      <td>4896</td>
      <td>-7</td>
      <td>42</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>496</td>
      <td>507</td>
      <td>+11</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage/contact)

### Misc

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fc4b0</th>
      <th colspan=4>#31268 ba5ae5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>64.83%</td>
      <td>63.16%</td>
      <td>-1.67%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>271</td>
      <td>264</td>
      <td>-7</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>147</td>
      <td>154</td>
      <td>+7</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage_diff/misc)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage/misc)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fc4b0</th>
      <th colspan=4>#31268 ba5ae5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.09%</td>
      <td>86.07%</td>
      <td>-0.02%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13786</td>
      <td>13783</td>
      <td>-3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2228</td>
      <td>2231</td>
      <td>+3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31268/coverage/phase_field)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31268/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 60.00% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Support aligned node processing when projecting primary nodes

lindsayad committed:
Store AMG and MortarData as pointers to reduce compilation deps

lindsayad committed:
Always treat displaced mortar mesh as replicated

lindsayad committed:
Don't bind reference to temporary

lindsayad committed:
Add debug_mesh as a parameter in the ContactAction

lindsayad committed:
Remove deprecated component parameter for ADPressure

lindsayad committed:
Make all WeightedGapUserObjects capable of computing contact pressure

Also rename PenaltyMortarUserObjectAux to MortarUserObjectAux since
this object was already capable of computing many quantities for
non-penalty based UOs

lindsayad committed:
Add gap and contact pressure for mortar in ContactAction

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3173453/), step Clang format on 879a5ab wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31268/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31268/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b36d438760c22efa38524993caa7cc672eff2c2e`

lindsayad committed:
Don't run block restricted sc fsp test in dbg

I've seen it timeout

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1136,7 +1136,14 @@ FEProblemBase::initialSetup()
           _neighbor_material_props.hasStatefulProperties())
+      {
         _has_initialized_stateful = true;
+        if (_ics.hasObjects() && _app.isRestarting())
```
I reverted that change

GiudGiud commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -634,47 +664,71 @@ ContactAction::act()
-    _problem->addAuxKernel("ContactPressureAux", name, params);
+      std::string name = _name + "_contact_pressure";
+      params.set<ExecFlagEnum>("execute_on",
+                               true) = {EXEC_NONLINEAR, EXEC_TIMESTEP_END, EXEC_TIMESTEP_BEGIN};
```
it s a lot of execution when you did not set a schedule for MortarUserObjectAux

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1858,6 +1861,105 @@ AutomaticMortarGeneration::projectSecondaryNodes()
+  // First, determine "on left" vs. "on right" orientation of the nodal neighbors.
+  // There can be a max of 2 nodal neighbors, and we want to make sure that the
+  // secondary nodal neighbor on the "left" is associated with the primary nodal
+  // neighbor on the "left" and similarly for the "right".
```
I don't know. Do you have copilot installed in vs code?

lindsayad commented on modules/contact/test/tests/multiple_contact_pairs/multiple_pairs_mortar.cmp:
```diff
@@ -0,0 +1,13 @@
+TIME STEPS relative 1.e-6 floor 0.0     # min:               0 @ t1 max:              18 @ t10
+
+NODAL VARIABLES relative 1.0e-5 floor 1.0e-8
+	disp_x                 # min:               0 @ t1,n1	max:      0.23836948 @ t10,n38
```
No, those are the min and max values of the field

GiudGiud replied:
this isnt used for anything right? Like it does not impose a criterion on the new min/max

lindsayad committed:
Address Guillaume review

- Typo fixes
- Document choice of replicated vs. distributed for mortar segment mesh
- Better method name: verifyLagrange

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3324696/), step Clang format on 50e3112 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31268/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31268/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5f1829d619a619865513d9a55cc16d23f6c15fc0`

lindsayad committed:
Draw diagram showing how to match sides of aligned nodes

lindsayad committed:
Fix bad use of secondary_node in place primary_node

Also use more const

GiudGiud commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1858,6 +1861,122 @@ AutomaticMortarGeneration::projectSecondaryNodes()
+  //  the cross product). Doing the same with the aligned primary node, if we pick the top primary
+  //  element, then the cross product also points out of the screen. Because the cross products
+  //  point in the same direction (positive dot product), then we know to associate the
+  //  secondary-primary pair. If we had picked the bottom primary element whose cross product points
```
```suggestion
  //  secondary-primary element pair. If we had picked the bottom primary element whose cross product points
```

GiudGiud commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1858,6 +1861,122 @@ AutomaticMortarGeneration::projectSecondaryNodes()
+
+  for (const auto nn : index_range(*secondary_node_neighbors))
+  {
+    const Elem * const secondary_neigh = (*secondary_node_neighbors)[nn];
```
```suggestion
    const Elem * const secondary_neigh = (*secondary_node_neighbors)[nn];
    mooseAssert(const secondary_neigh->first_order_equivalent_type() == EDGE2, "Should be an edge");
```
This routine is only in the context of 2D mortar right?

GiudGiud commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1858,6 +1861,122 @@ AutomaticMortarGeneration::projectSecondaryNodes()
+  }
+
+  // We need to handle the case where we've exactly projected a secondary node onto a
+  // primary node, but our secondary node is at one of the secondary face endpoints and
```
```suggestion
  // primary node, but our secondary node is at one of the secondary boundary face endpoints and
```

GiudGiud commented on modules/contact/src/userobjects/LMWeightedGapUserObject.C:
```diff
@@ -58,6 +57,26 @@ LMWeightedGapUserObject::LMWeightedGapUserObject(const InputParameters & paramet
 
+void
+LMWeightedGapUserObject::checkInput(const MooseVariable * const var,
+                                    const std::string & var_name) const
```
```suggestion
                                    const std::string & var_param_name) const
```

lindsayad committed:
Address Guillaume second review

- Rename MortarData -> MortarInterfaceWarehouse
- More descriptive wording in AutomaticMortarGeneration
- Better method parameter names
